### PR TITLE
(refactor) Changed scope of EngineInfo structure from Global to Local

### PR DESCRIPTION
**Pull request by rahulchheda**
Signed-off-by: Rahul M Chheda <rahul.chheda@mayadata.io>

**What this PR does / why we need it**:

- In this PR, we have converted the global EngineInfo variable, to local variable.
- There was an issue in chaos-operator where the client-go k8s api-server calls to perform CRUD operations on ChaosEngine, sometimes failed and caused chaos-operator error logs.
- This was because of change in resourceVersion of ChaosEngine resource, because both the Global Variable of EngineInfo, and local instances of EngineInfo were not in sync before make k8s client-go operations.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Checklist:**
- [ ] Fixes #<issue number>
- [ ] Labelled this PR & related issue with `documentation` tag
- [ ] PR messages has document related information
- [ ] Labelled this PR & related issue with `breaking-changes` tag
- [ ] PR messages has breaking changes related information
- [ ] Labelled this PR & related issue with `requires-upgrade` tag
- [ ] PR messages has upgrade related information
- [ ] Commit has unit tests
- [ ] Commit has integration tests

### CODE REVIEW

File: pkg/controller/chaosengine/chaosengine_controller.go [hunk 1]
```diff
@@ -125,8 +123,8 @@ func watchChaosResources(clientSet client.Client, c controller.Controller) error
 // Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
 func (r *ReconcileChaosEngine) Reconcile(request reconcile.Request) (reconcile.Result, error) {
 	reqLogger := startReqLogger(request)
-
-	err := r.getChaosEngineInstance(request)
+	engine := InitEngineInfo()
```
Comment (chandankumar4) [inline pkg/controller/chaosengine/chaosengine_controller.go:126]: Replace this function with var

File: tests/bdd/bdd_test.go [hunk 2]
```diff
@@ -389,6 +392,46 @@ var _ = Describe("BDD on chaos-operator", func() {
...
+			)
+
+			fmt.Printf("Got Pod Name: %v\n", podName)
+			podLogOpts := v1.PodLogOptions{}
+			req := client.CoreV1().Pods("litmus").GetLogs(podName, &podLogOpts)
+			podLogs, err := req.Stream()
+			Expect(err).To(BeNil())
+			// if err != nil {
+			// 	return "error in opening stream"
+			// }
```
Comment (rajdas98) [inline tests/bdd/bdd_test.go:419]: Can u remove this ?

File: tests/bdd/bdd_test.go [hunk 3]
```diff
@@ -389,6 +392,45 @@ var _ = Describe("BDD on chaos-operator", func() {
...
+				if matched == true {
+					podName = pods.Items[i].Name
+				}
+			}
+			Expect(podName).To(
+				Not(BeEmpty()),
+				"Unable to get the pod, might be something wrong with chaos-operator BDD",
+			)
+
+			fmt.Printf("Got Pod Name: %v\n", podName)
```
Comment (chandankumar4) [inline tests/bdd/bdd_test.go:412]: Can we use the klog here also

Comment (chandankumar4) [reply tests/bdd/bdd_test.go:412]: Check other places also

File: tests/bdd/bdd_test.go [hunk 4]
```diff
@@ -389,12 +392,78 @@ var _ = Describe("BDD on chaos-operator", func() {
...
 
 //Deleting all unused resources
 var _ = AfterSuite(func() {
 
-	By("Deleting all CRDs")
-	crdDeletion := exec.Command("kubectl", "delete", "-f", "../../deploy/chaos_crds.yaml").Run()
-	Expect(crdDeletion).To(BeNil())
+	//Deleting ChaosExperiments
+	By("Deleting ChaosExperiments")
+	err = exec.Command("kubectl", "delete", "chaosexperiments", "--all", "-n", "litmus").Run()
```
Comment (chandankumar4) [inline tests/bdd/bdd_test.go:447]:
```suggestion
	err := exec.Command("kubectl", "delete", "chaosexperiments", "--all", "-n", "litmus").Run()
```

File: pkg/controller/chaosengine/chaosengine_controller.go [hunk 5]
```diff
@@ -421,13 +417,14 @@ func getAnnotationCheck() error {
...
 // reconcileForDelete reconciles for deletion/force deletion of Chaos Engine
-func (r *ReconcileChaosEngine) reconcileForDelete(request reconcile.Request) (reconcile.Result, error) {
+func (r *ReconcileChaosEngine) reconcileForDelete(engine *chaosTypes.EngineInfo, request reconcile.Request) (reconcile.Result, error) {
 	err := r.forceRemoveChaosResources(engine, request)
 	if err != nil {
 		r.recorder.Eventf(engine.Instance, corev1.EventTypeWarning, "ChaosResourcesOperationFailed", "Unable to delete chaos resources")
 		return reconcile.Result{}, err
 	}
-	opts := client.UpdateOptions{}
+	//opts := client.UpdateOptions{}
```
Comment (chandankumar4) [inline pkg/controller/chaosengine/chaosengine_controller.go:426]: remove this line

Comment (rahulchheda) [reply pkg/controller/chaosengine/chaosengine_controller.go:426]: Got it.

File: tests/bdd/bdd_test.go [hunk 6]
```diff
@@ -389,12 +392,79 @@ var _ = Describe("BDD on chaos-operator", func() {
...
+
+	//Deleting rbacs
+	By("Deleting RBAC's")
+	err = exec.Command("kubectl", "delete", "-f", "../../deploy/rbac.yaml").Run()
+	Expect(err).To(BeNil())
+
+	//Deleting Chaos-Operator
+	By("Deleting operator")
+	err = exec.Command("kubectl", "delete", "-f", "../../deploy/operator.yaml").Run()
+	Expect(err).To(BeNil())
```
Comment (chandankumar4) [inline tests/bdd/bdd_test.go:468]: Remove it in reverse way like: 
1 - operator.yaml 
2 - rbac.yaml
3 - chaos_crds.yaml

Comment (rahulchheda) [reply tests/bdd/bdd_test.go]: Fixed @chandankumar4

Comment (rahulchheda) [reply tests/bdd/bdd_test.go]: Take a look now, `rbac.yaml` was deleting the namespace `litmus`, thats why operator deletion was failing

**Pull request merged**